File: node/test/smoke.js
'use strict';

const assert = require('node:assert/strict');

const Taskflow = require('..');

async function main() {
  const orchestrator = new Taskflow.Orchestrator({ memoryResultStorage: true });

  try {
    orchestrator.registerTask('fetch_order', (ctx) => {
      ctx.set('orderId', 'order-42');
      ctx.setResult('payload', 'order-42');
      return Taskflow.constants.TF_TASK_STATE_SUCCESS;
    });

    orchestrator.registerTask('send_email', (ctx) => {
      const payload = ctx.getResult(1n, 'payload');
      assert.equal(payload, 'order-42');
      ctx.set('emailStatus', 'sent');
      return 'success';
    });

    const blueprintId = 1n;
    const blueprint = new Taskflow.Blueprint()
      .addNode({ id: 1, taskType: 'fetch_order', label: 'Fetch Order' })
      .addNode({ id: 2, taskType: 'send_email', label: 'Send Email' })
      .addEdge({ from: 1, to: 2 });

    console.log('=== TaskFlow Node.js Smoke Example ===');
    console.log('orchestrator handle:', orchestrator.handle.toString());
    console.log('register blueprint:', blueprintId.toString());
    console.log(blueprint.toJSON());

    orchestrator.registerBlueprint(blueprintId, blueprint);
    blueprint.destroy();

    const execution = orchestrator.runAsync(blueprintId);
    try {
      const finalState = await execution.wait();
      const snapshot = execution.snapshot();
      const nodeStateMap = new Map(snapshot.nodes.map((node) => [node.id, node]));

      assert.equal(finalState, Taskflow.constants.TF_TASK_STATE_SUCCESS);
      assert.equal(execution.isComplete(), true);
      assert.equal(execution.getOverallState(), Taskflow.constants.TF_TASK_STATE_SUCCESS);
      assert.equal(execution.getNodeState(1n), Taskflow.constants.TF_TASK_STATE_SUCCESS);
      assert.equal(execution.getNodeState(2n), Taskflow.constants.TF_TASK_STATE_SUCCESS);
      assert.equal(execution.get('orderId'), 'order-42');
      assert.equal(execution.get('emailStatus'), 'sent');
      assert.equal(execution.getResult(1n, 'payload'), 'order-42');
      assert.equal(nodeStateMap.get(1)?.state, 'success');
      assert.equal(nodeStateMap.get(2)?.state, 'success');

      console.log('execution handle:', execution.handle.toString());
      console.log('execution id:', execution.id().toString());
      console.log('final state:', Taskflow.native.taskStateName(finalState));
      console.log('snapshot:');
      console.log(JSON.stringify(snapshot, null, 2));
    } finally {
      execution.destroy();
      console.log('execution destroyed');
    }
  } finally {
    orchestrator.destroy();
    console.log('orchestrator destroyed');
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});

File: Python/test/smoke.py
from __future__ import annotations

import importlib.util
from pathlib import Path


def load_taskflow_module():
    root = Path(__file__).resolve().parents[2]
    candidates = list((root / "build" / "Python").glob("taskflow.cpython-*.so"))
    candidates += list((root / "build" / "Python").glob("taskflow*.pyd"))

    if not candidates:
        raise RuntimeError("taskflow Python extension not found; build the `taskflow_python` target first")

    module_path = candidates[0]
    spec = importlib.util.spec_from_file_location("taskflow", module_path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"failed to load module spec from {module_path}")

    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


taskflow = load_taskflow_module()


def main() -> None:
    orchestrator = taskflow.Orchestrator({"memory_result_storage": True})

    try:
        def fetch_order(ctx):
            assert ctx.exec_id() > 0
            assert ctx.node_id() == 1
            ctx.set("order_id", "order-42")
            ctx.set_result("payload", "order-42")
            return taskflow.constants.TF_TASK_STATE_SUCCESS

        def send_email(ctx):
            assert ctx.exec_id() > 0
            assert ctx.node_id() == 2
            assert ctx.get("order_id") == "order-42"
            assert ctx.get_result(1, "payload") == "order-42"
            ctx.set("email_status", "sent")
            return "success"

        orchestrator.register_task("fetch_order", fetch_order)
        orchestrator.register_task("send_email", send_email)

        blueprint = taskflow.Blueprint()
        blueprint.add_node({"id": 1, "task_type": "fetch_order", "label": "Fetch Order"})
        blueprint.add_node({"id": 2, "task_type": "send_email", "label": "Send Email"})
        blueprint.add_edge({"from": 1, "to": 2})

        orchestrator.register_blueprint(1, blueprint)
        blueprint.destroy()

        execution = orchestrator.run_sync(1)
        try:
            assert execution.get_overall_state() == taskflow.constants.TF_TASK_STATE_SUCCESS
            assert execution.get_node_state(1) == taskflow.constants.TF_TASK_STATE_SUCCESS
            assert execution.get_node_state(2) == taskflow.constants.TF_TASK_STATE_SUCCESS
            assert execution.get("order_id") == "order-42"
            assert execution.get("email_status") == "sent"
            assert execution.get_result(1, "payload") == "order-42"

            snapshot = execution.snapshot()
            print("=== TaskFlow Python Smoke Example ===")
            print("orchestrator handle:", orchestrator.handle)
            print("execution handle:", execution.handle)
            print("execution id:", execution.id())
            print("final state:", taskflow.task_state_name(execution.get_overall_state()))
            print("snapshot:", snapshot)
        finally:
            execution.destroy()
    finally:
        orchestrator.destroy()


if __name__ == "__main__":
    main()
